File: Graphs/DFSGraphs/source/dfs_graph.py
from collections import defaultdict


class DFSGraph:
    def __init__(self):
        """
        Initialize the graph with an empty adjacency list.

        Attributes:
            graph (defaultDict): A dictionary where each key is a vertex and the value is a list of adjacent vertices.
            vertices (set): A set containing all vertices in the graph.
        """
        self.graph = defaultdict(list)
        self.vertices = set()

    def add_vertex(self, vertex):
        """
        Add a vertex to the graph.

        Args:
            vertex (str): The vertex to be added.
        """
        self.vertices.add(vertex)

    def add_edge(self, u, v):
        """
        Add a directed edge from vertex u to vertex v. If the vertices do not exist, they are added to the graph.

        Args:
            u (str): The starting vertex of the edge.
            v (str): The ending vertex of the edge.

        Prints:
            str: Confirmation message that the edge was added successfully.
        """
        if u not in self.vertices:
            self.add_vertex(u)
        if v not in self.vertices:
            self.add_vertex(v)

        self.graph[u].append(v)
        print(f"Edge {u} -> {v} added successfully.")

    def dfs(self, start_vertex):
        """
        Perform a Depth-First Search (DFS) traversal starting from the specified vertex.

        Args:
            start_vertex (str): The vertex from which to start the DFS traversal.

        Prints:
            str: The vertices visited during the DFS traversal in the order they are visited.
        """
        visited = set()

        def dfs_recursive(vertex):
            if vertex not in visited:
                visited.add(vertex)
                print(vertex, end=" ")
                for neighbor in self.graph[vertex]:
                    dfs_recursive(neighbor)

        dfs_recursive(start_vertex)
        print()

    def display(self):
        """
        Display the graph as an adjacency list.

        Prints:
            str: Each vertex and its list of adjacent vertices.
        """
        for vertex in self.vertices:
            print(f"{vertex} -> {self.graph[vertex]}")

File: Graphs/DFSGraphs/source/dfs_graph.js
/**
 * Class representing a directed graph with Depth-First Search (DFS) functionality.
 */
class DFSGraph {
    /**
     * Initialize the graph with an empty adjacency list and a set of vertices.
     *
     * @property {Object} graph - A dictionary where each key is a vertex, and the value is an array of adjacent vertices.
     * @property {Set} vertices - A set containing all vertices in the graph.
     */
    constructor() {
        this.graph = {};
        this.vertices = new Set();
    }

    /**
     * Add a vertex to the graph.
     *
     * @param {string} vertex - The vertex to be added.
     */
    addVertex(vertex) {
        this.vertices.add(vertex);
    }

    /**
     * Add a directed edge from the source vertex to the target vertex.
     * If the vertices do not exist, they are added to the graph.
     *
     * @param {string} source - The starting vertex of the edge.
     * @param {string} target - The ending vertex of the edge.
     */
    addEdge(source, target) {
        if (!this.vertices.has(source)) {
            this.addVertex(source);
        }
        if (!this.vertices.has(target)) {
            this.addVertex(target);
        }

        if (!this.graph[source]) {
            this.graph[source] = [];
        }
        this.graph[source].push(target);
    }

    /**
     * Perform a Depth-First Search (DFS) traversal starting from the specified vertex.
     * Uses an iterative approach with a stack to avoid recursion.
     *
     * @param {string} startVertex - The vertex from which to start the DFS traversal.
     * @returns {void}
     *
     * @example
     * graph.dfs("A");
     * // Output: A B D E F C
     */
    dfs(startVertex) {
        const visited = new Set();
        const stack = [startVertex];
        const result = []; // Array to collect the visited vertices

        while (stack.length > 0) {
            const currentVertex = stack.pop();
            if (!visited.has(currentVertex)) {
                visited.add(currentVertex);
                result.push(currentVertex); // Collect the vertex

                // Add neighbors in reverse order to maintain order of traversal
                for (let i = (this.graph[currentVertex] || []).length - 1; i >= 0; i--) {
                    const neighbor = this.graph[currentVertex][i];
                    if (!visited.has(neighbor)) {
                        stack.push(neighbor);
                    }
                }
            }
        }

        // Print all the visited vertices in one row
        console.log(result.join(" "));
    }

    /**
     * Display the graph as an adjacency list.
     *
     * @returns {void}
     *
     * @example
     * graph.display();
     * // Output:
     * // A -> B, C
     * // B -> D
     * // C -> D, F
     */
    display() {
        for (const vertex of this.vertices) {
            console.log(`${vertex} -> ${this.graph[vertex] ? this.graph[vertex].join(", ") : ""}`);
        }
    }
}

module.exports = DFSGraph;
